sapi/cli/php_cli_server.c

```diff
@@ -1680,14 +1680,37 @@ static void php_cli_server_client_save_header(php_cli_server_client *client)
 {
 	/* Wrap header value in a zval to add is to the HashTable which acts as an array */
 	zval tmp;
-	ZVAL_STR(&tmp, client->current_header_value);
 	/* strip off the colon */
 	zend_string *lc_header_name = zend_string_tolower_ex(client->current_header_name, /* persistent */ true);
 	GC_MAKE_PERSISTENT_LOCAL(lc_header_name);
 
-	/* Add the wrapped zend_string to the HashTable */
-	zend_hash_add(&client->request.headers, lc_header_name, &tmp);
-	zend_hash_add(&client->request.headers_original_case, client->current_header_name, &tmp);
+	zval *entry = zend_hash_find(&client->request.headers, lc_header_name);
+	bool with_comma = !zend_string_equals_literal(lc_header_name, "set-cookie");
+
+	/**
+	 * `Set-Cookie` HTTP header being the exception, they can have 1 or more values separated
+	 * by a comma while still possibly be set separately by the client.
+	 **/
+	if ((entry == NULL && with_comma) || !with_comma) {
+		ZVAL_STR(&tmp, client->current_header_value);
+
+		/* Add the wrapped zend_string to the HashTable */
+		zend_hash_add(&client->request.headers, lc_header_name, &tmp);
+		zend_hash_add(&client->request.headers_original_case, client->current_header_name, &tmp);
+	} else {
+		zend_string *curval = Z_STR_P(entry);
+		zend_string *newval = zend_string_alloc(ZSTR_LEN(curval) + ZSTR_LEN(client->current_header_value) + 3, /* persistent */true);
```

Comment:
Please use `zend_string_safe_alloc` instead when adding two string lengths together unless you're adding a small constant or are certain that the sum cannot overflow ZSTR_MAX_LEN under any condition.

Reply:
That's actually a great suggestion.